### DISCUSSION TITLE
Mirth 4.5.0 removes &#xD; encoding in Webservice call

### DISCUSSION THREAD

**Discussion by MirthRanger** [nextgenhealthcare/connect#6200] [Q&A]
We have just upgraded to 4.5.0 and found that the &#xD; encoding in the Webservice SOAP body is being replaced with the 0x0d character when written out the TCP stream - as seen by a wireshark capture.  It didn't' happen in 4.4.2.  Has anyone else seen this?

**Comment by medir** [comment]
We have detected the same issue and also noticed that in version <= 4.4.2, HL7 responses encapsulated in XML (Web Service Listener) had the segment code (CR) represented as `&#xd;`. Starting from version >= 4.5.0, we receive a different encoding for CR `0x0D`.

We are attaching the files where this change in encoding can be seen:
[output_mc_v451.txt](https://github.com/user-attachments/files/17061545/output_mc_v451.txt)
[output_mc_v442.txt](https://github.com/user-attachments/files/17061546/output_mc_v442.txt)

**Reply by tonygermano** [reply]
This does look like a bug. What I saw when trying to parse the resulting xml, and then confirmed with [this SO post](https://stackoverflow.com/a/2266166), the literal carriage return gets converted to a linefeed, making the hl7 no longer valid. So, it does need to be encoded (either as `&#xd;` or `&#13;`) to retain the same value when parsing.

**Comment by mrreview** [comment]
We're also facing this exact problem.

**Reply by nicovn** [reply]
There is indeed a difference, but I don't have any issues parsing the HL7 ACK message in the response transformer in Mirth...

`// Use HL7 ACK from SOAP response envelope

var hl7ACKMessage = msg.*::Body.*::acceptMessageResponse.*.toString();


// Convert the HL7 ACK (string) to a Mirth HL7 v2 XML representation for further processing
msg = SerializerFactory.getSerializer('HL7V2').toXML(hl7ACKMessage);`


Bet Regards

Nico

Check out our latest and free FHIR search tool: FHINDR - [https://github.com/amaron-team/fhindr](https://github.com/amaron-team/fhindr)

![to_mirth_4_5_0](https://github.com/user-attachments/assets/050be9be-37d2-444c-ac06-e8b37a7642cc)
![to_mirth_4_3_0](https://github.com/user-attachments/assets/2e026da8-e798-4dd6-8596-2a032159517f)
![ack-message](https://github.com/user-attachments/assets/c81b4446-bbfb-4fa3-9a2d-a8371927b658)

**Reply by MirthRanger** [reply]
The issue we are encountering is for Webservice to our EMR.  We are sending an encoded HL7 V2.x via SOAP webservice.  Whilst we are sending the segment delimiter as encoded <CR> ie &#xD; the mirth webservice sending is decoding it to 0x0d; as captured with TCP dump.  This shouldn't happen.  When our EMR receives, per the XML standard 0x0d is returned as 0x0a and so the HL7 encoding of the message in the payload is broken.  Nextgen have confirmed this is a problem and planning to fix in 4.6.0.

Thank you for investigating,

Mark

**Reply by medir** [reply]
Hello @MirthRanger, do you have a link of the issue you mention? I would like to track it.

> Nextgen have confirmed this is a problem and planning to fix in 4.6.0.

Thank you.

**Comment by nicovn** [comment]
Hi Mark,

FYI - the segment separator sent via the Web Service Sender is different depending on how you encode the message in the template in Mirth.

e.g.

`<?xml version="1.0" encoding="UTF-8"?><SOAP-ENV:Envelope xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/" xmlns:tns="http://ws.connectors.connect.mirth.com/">
  <SOAP-ENV:Header/>
  <SOAP-ENV:Body>
    <tns:acceptMessage>
      <arg0><![CDATA[${message.encodedData}]]></arg0>
    </tns:acceptMessage>
  </SOAP-ENV:Body>
</SOAP-ENV:Envelope>
`


-> the above will send line feed (0A) as segment character

`<?xml version="1.0" encoding="UTF-8"?><SOAP-ENV:Envelope xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/" xmlns:tns="http://ws.connectors.connect.mirth.com/">
  <SOAP-ENV:Header/>
  <SOAP-ENV:Body>
    <tns:acceptMessage>
      <arg0>${XmlUtil.encode(${message.encodedData})}</arg0>
    </tns:acceptMessage>
  </SOAP-ENV:Body>
</SOAP-ENV:Envelope>
`

-> if using XmlUtil.encode the segment separator will be carriage return (0D)

![segment-separator](https://github.com/user-attachments/assets/87d0b9bd-ec0c-4a8a-a5e4-4283d4e43c15)


Best Regards

Nico

Check out our latest and free FHIR search tool: FHINDR - [](https://github.com/amaron-team/fhindr)

**Reply by MirthRanger** [reply]
Hi Nico,

Thank you for investigating further.  I am doing the XML encoding in the outbound, which is encoding with the expected XML encoding &#13; .  This can be seen in the Sent message window in Mirth administrator, for example:

![image](https://github.com/user-attachments/assets/0a7a2a60-ea3a-4b0a-a2c2-943a9c540445)

The issue seems to be that when Mirth writes this out the TCP stream it is reverting it back to 0d, as you see in your packet capture.  We should be seeing &#13; at the position of segment delimiters in the capture.  Seems like the Webservice sender TCP writer is decoding the XML in the payload prior to sending to the endpoint - it should be leaving the encoding in place.  

Our problem then arises at our EMR, which appears to be interpreting the XML to the standard:

https://www.w3.org/TR/REC-xml/#sec-line-ends
=================
2.11 End-of-Line Handling
XML parsed entities are often stored in computer files which, for editing convenience, are organized into lines. These lines are typically separated by some combination of the characters CARRIAGE RETURN (#xD) and LINE FEED (#xA).

To simplify the tasks of applications, the XML processor MUST behave as if it normalized all line breaks in external parsed entities (including the document entity) on input, before parsing, by translating both the two-character sequence #xD #xA and any #xD that is not followed by #xA to a single #xA character.


=================

So our EMR is reading in 0a and that is breaking the HL7 message payload in the XML.  Based on packet captures I have done this seemed to start in 4.5.0

Cheers, Mark

**Comment by MirthRanger** [comment]
Hello, to close this discussion, Nextgen's resolution to this is to reinstall 2 x JAR files removed in the 4.5.0 upgrade -  stax2-api-4.2.2 and woodstox-core-7.1.0.  Note, these had been removed due to security concerns and will not be replaced in future releases.  Mark
